packages/react/src/components/dialog/Dialog.tsx

```diff
@@ -0,0 +1,144 @@
+  'aria-describedby'?: string;
+  /**
+   * When `true`, dialog is visible,
+   */
+  isOpen: boolean;
+  /**
+   * A function to control the isOpen property.
+   */
+  close?: () => void;
+  /**
+   * The id of the element which will get focus after the dialog is closed.
+   */
+  focusAfterCloseId?: string;
+  /**
+   * Custom theme styles
+   */
+  theme?: DialogCustomTheme;
+  /**
+   * Additional class names to apply to the dialog.
+   */
+  className?: string;
+  /**
+   * Children to render inside the dialog element.
+   */
+  children: ReactNode | ReactNodeArray;
+}>;
+
+export const Dialog = ({ id, isOpen, children, close, className, focusAfterCloseId, theme, ...props }: DialogProps) => {
+  const customThemeClass = useTheme<DialogCustomTheme>(styles.dialogContainer, theme);
+  const dialogRef: RefObject<HTMLInputElement> = React.createRef();
+
+  const { 'aria-labelledby': ariaLabelledby, 'aria-describedby': ariaDescribedby } = props;
+
+  const onKeyDown = (event: KeyboardEvent): void => {
+    if (event.key === 'Escape') {
+      close();
+    }
+  };
+
+  const setFocusAfterClose = (): void => {
+    const focusElement = document.getElementById(focusAfterCloseId);
+    if (focusElement) {
+      focusElement.focus();
+    } else {
+      // eslint-disable-next-line no-console
+      console.warn('The "focusAfterCloseId" property did not match any element.');
+    }
+  };
+
+  useEffect(() => {
+    if (isOpen) {
+      document.body.classList.add(styles.dialogVisibleBody);
+      document.addEventListener('keydown', onKeyDown, false);
+    }
+
+    return (): void => {
+      if (isOpen) {
+        document.removeEventListener('keydown', onKeyDown, false);
+        document.body.classList.remove(styles.dialogVisibleBody);
+        if (focusAfterCloseId) {
+          setFocusAfterClose();
+        }
+      }
+    };
+  });
+
+  const onTabBarrierFocus = (isFirst: boolean) => () => {
+    if (dialogRef.current) {
+      const focusableElements: NodeList = dialogRef.current.querySelectorAll(
+        'a, button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select',
+      );
+
+      if (focusableElements.length) {
+        if (isFirst) {
+          (focusableElements[focusableElements.length - 1] as HTMLElement).focus();
+        } else {
+          (focusableElements[0] as HTMLElement).focus();
+        }
+      }
+    }
+  };
+
+  const DialogComponent = (): JSX.Element => (
+    <div className={classNames(styles.dialogContainer, customThemeClass)}>
+      <div className={styles.dialogBackdrop} />
+      <div
+        ref={dialogRef}
+        role="dialog"
+        aria-modal="true"
+        id={id}
+        className={classNames(styles.dialog, className)}
+        aria-labelledby={ariaLabelledby}
+      >
+        <div aria-describedby={ariaDescribedby}>
+          {/* eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex */}
+          <div onFocus={onTabBarrierFocus(true)} tabIndex={0} />
```

Comment:
What determines the initially focused element?

Reply:
It is now fixed in the Dialog.Header but I could get rid of the Header component. 
Then the focus handling would be in one place.